client/hostvolumemanager/host_volume_plugin.go

```diff
@@ -110,20 +124,65 @@ func (p *HostVolumePluginMkdir) Create(_ context.Context,
 		return resp, nil
 	} else if !os.IsNotExist(err) {
 		// doesn't exist, but some other path error
-		log.Debug("error with plugin", "error", err)
+		log.Debug("error with path", "error", err)
 		return nil, err
 	}
 
-	err := os.MkdirAll(path, 0o700)
+	params, err := decodeMkdirParams(req.Parameters)
 	if err != nil {
-		log.Debug("error with plugin", "error", err)
+		log.Error("error with parameters", "error", err)
 		return nil, err
 	}
 
+	err = os.MkdirAll(path, params.Mode)
+	if err != nil {
+		log.Error("error creating directory", "error", err)
+		return nil, fmt.Errorf("error creating directory: %w", err)
+	}
+
+	// Chown note: A uid or gid of -1 means to not change that value.
+	if err = os.Chown(path, params.Uid, params.Gid); err != nil {
+		log.Error("error changing owner/group", "error", err, "uid", params.Uid, "gid", params.Gid)
+		return nil, fmt.Errorf("error changing owner/group: %w", err)
+	}
+
 	log.Debug("plugin ran successfully")
 	return resp, nil
 }
 
+func decodeMkdirParams(in map[string]string) (HostVolumePluginMkdirParams, error) {
+	var out HostVolumePluginMkdirParams
+
+	var meta mapstructure.Metadata
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result:           &out,
+		Metadata:         &meta,
+		ErrorUnused:      true, // error on unexpected config fields
+		WeaklyTypedInput: true, // convert strings to target types
+	})
+	if err != nil {
+		return out, fmt.Errorf("error creating decoder: %w", err)
+	}
+	err = decoder.Decode(in)
+	if err != nil {
+		return out, fmt.Errorf("error decoding: %w", err)
+	}
```

Comment:
This all seems kinda heavyweight given that we're not doing any serialization with it. We have only 3 fields with well-known conversions, and there's no reuse of this decoder struct. We can do all of this with a couple of function calls instead and get more application-specific errors than "decoding" errors.